assets/app/features/upgrade/upgrade.controller.spec.js

```diff
@@ -44,14 +54,51 @@ describe('Upgrade Controller', function () {
             should.exist(controller.steps);
         });
 
-        it('should contains a nextStep function ', function() {
-            should.exist(controller.steps.nextStep);
-            expect(controller.steps.nextStep).toEqual(jasmine.any(Function));
+        describe('nextStep function', function () {
+            it('should exist', function () {
+                should.exist(controller.steps.nextStep);
+                expect(controller.steps.nextStep).toEqual(jasmine.any(Function));
+            });
+
+            it('when called it should move to the next step', function () {
+                expect($state.current.id).toBe(0);
+                // calculate the next step in the list
+                var nextStep = (controller.steps.list[upgradeStepsFactory.activeStep.id + 1]).state;
+                controller.steps.nextStep();
+                expect($state.go).toHaveBeenCalledWith(nextStep);
+            });
+
+            it('when called in the last step it should not change steps', function () {
+                // set the last step in the list as current
+                $state.current = {id: 0, name: 'upgrade.upgrade-nodes', state: 'upgrade.upgrade-nodes'};
+                // refresh the list so it sets it as active
+                upgradeStepsFactory.refeshStepsList();
```

Comment:
Such problem should be validated on the service unit test, not this. Is not a responsibility of this controller to refresh the list of steps.
We need to enforce as many test scenarios as possible to make our test to fail if something goes wrong, but just to ensure each function does what is supposed to.

I think the way things are integrated and supposed to be working all together will be covered by automation tests.